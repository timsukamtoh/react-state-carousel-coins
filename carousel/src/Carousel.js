import { useState } from "react";
import "./Carousel.css";
import Card from "./Card";


/** Carousel: displays images and arrows to navigate through them
 *
 * Props:
 * - photos: array of {src, caption} objects
 * - title: string describing the collection of images
 *
 * State:
 * - currCardIdx: integer for current card index
 *
 * App --> Carousel --> Card
 */
 function Carousel({ photos, title }) {
  const [currCardIdx, setCurrCardIdx] = useState(0);

  const currCard = photos[currCardIdx];
  const total = photos.length;

  //Increments currCardIdx state by 1
  function goForward() {
    setCurrCardIdx((currCardIdx + 1)%photos.length);
  }

  //Increments currCardIdx state by 1
  function goBackward() {
    let newIdx = currCardIdx < 1 ? photos.length-1: currCardIdx - 1;

    setCurrCardIdx(newIdx);
  }

  return (
    <div className="Carousel">
      <h1>{title}</h1>
      <div className="Carousel-main">
        <i
          className="bi bi-arrow-left-circle"
          onClick={goBackward}
        />
        <Card
          caption={currCard.caption}
          src={currCard.src}
          currNum={currCardIdx + 1}
          totalNum={total}
        />
        <i
          className="bi bi-arrow-right-circle"
          onClick={goForward}
        />
      </div>
    </div>
  );
}

export default Carousel;
