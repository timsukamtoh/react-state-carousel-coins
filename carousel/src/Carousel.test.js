import { render, fireEvent } from "@testing-library/react";
import Carousel from "./Carousel";
import TEST_IMAGES from "./_testCommon.js";


describe("carousel arrows work", function () {
  it("works when you click on the right arrow", function () {
    const { container } = render(
      <Carousel
        photos={TEST_IMAGES}
        title="images for testing"
      />
    );
    // expect the first image to show, but not the second
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).not.toBeInTheDocument();

    // move forward in the carousel
    const rightArrow = container.querySelector(".bi-arrow-right-circle");
    fireEvent.click(rightArrow);

    // expect the second image to show, but not the first
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).not.toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).toBeInTheDocument();
  });


  it("works when you click on the left arrow", function () {
    const { container, debug } = render(
      <Carousel
        photos={TEST_IMAGES}
        title="images for testing"
      />
    );
    // expect the first image to show, but not the second
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).not.toBeInTheDocument();

    // move forward in the carousel
    const rightArrow = container.querySelector(".bi-arrow-right-circle");
    fireEvent.click(rightArrow);

    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).not.toBeInTheDocument();

    // move backward in the carousel
    const leftArrow = container.querySelector(".bi-arrow-left-circle");
    fireEvent.click(leftArrow);

    // expect the second image to show, but not the first
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).not.toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).toBeInTheDocument();
  });

  it("loops forward", function () {
    const { container, debug } = render(
      <Carousel
        photos={TEST_IMAGES}
        title="images for testing"
      />
    );
    // expect the first image to show, but not the second
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).not.toBeInTheDocument();

    // move forward in the carousel
    const rightArrow = container.querySelector(".bi-arrow-right-circle");
    fireEvent.click(rightArrow);
    fireEvent.click(rightArrow);
    fireEvent.click(rightArrow);

    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 3"]')
    ).not.toBeInTheDocument();
  });

  it("loops backward", function () {
    const { container, debug } = render(
      <Carousel
        photos={TEST_IMAGES}
        title="images for testing"
      />
    );
    // expect the first image to show, but not the second
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).not.toBeInTheDocument();

    // move backward in the carousel
    const leftArrow = container.querySelector(".bi-arrow-left-circle");
    fireEvent.click(leftArrow);

    expect(
      container.querySelector('img[alt="testing image 3"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).not.toBeInTheDocument();
  });
});

describe("smoke tests", function () {
  it("renders carousel without crashing", function () {
    // this is a low-value test, but better than nothing
    render(<Carousel photos={[{}]} title="test" />);
  });
});


describe("snapshot tests", function () {
  it("initial snapshot", function () {
    const { container } = render(<Carousel photos={TEST_IMAGES} title="test" />);

    expect(container).toMatchSnapshot();
  });

  it("second snapshot", function () {
    const { container } = render(<Carousel photos={TEST_IMAGES} title="test" />);
    const rightArrow = container.querySelector(".bi-arrow-right-circle");
    fireEvent.click(rightArrow);

    expect(container).toMatchSnapshot();
  });


});


