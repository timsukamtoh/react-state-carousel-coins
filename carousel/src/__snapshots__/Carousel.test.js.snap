// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests initial snapshot 1`] = `
<div>
  <div
    class="Carousel"
  >
    <h1>
      test
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="bi bi-arrow-left-circle"
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          testing image 1
        </h4>
        <img
          alt="testing image 1"
          class="Card-image"
          src="test1.com"
        />
        <small
          class="Card-small"
        >
          Image 
          1
           of 
          3
          .
        </small>
      </div>
      <i
        class="bi bi-arrow-right-circle"
      />
    </div>
  </div>
</div>
`;

exports[`snapshot tests second snapshot 1`] = `
<div>
  <div
    class="Carousel"
  >
    <h1>
      test
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="bi bi-arrow-left-circle"
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          testing image 2
        </h4>
        <img
          alt="testing image 2"
          class="Card-image"
          src="test2.com"
        />
        <small
          class="Card-small"
        >
          Image 
          2
           of 
          3
          .
        </small>
      </div>
      <i
        class="bi bi-arrow-right-circle"
      />
    </div>
  </div>
</div>
`;
